src/array_string.rs

```diff
@@ -0,0 +1,186 @@
+/// though it would be safe to do so.
+#[derive(Copy)]
+pub struct ArrayString<A: Array<Item=u8>> {
+    xs: A,
+    len: A::Index,
+}
+
+unsafe fn new_array<A: Array<Item=u8>>() -> A {
+    // Note: Returning an uninitialized value here only works
+    // if we can be sure the data is never used. The nullable pointer
+    // inside enum optimization conflicts with this this for example,
+    // so we need to be extra careful. See `NoDrop` enum.
+    mem::uninitialized()
+}
+
+impl<A: Array<Item=u8>> ArrayString<A> {
+    /// Create a new empty `ArrayString`.
+    ///
+    /// Capacity is inferred from the type parameter.
+    ///
+    /// ```
+    /// use arrayvec::ArrayString;
+    ///
+    /// let mut string = ArrayString::<[_; 16]>::new();
+    /// string.push_str("foo");
+    /// assert_eq!(&string[..], "foo");
+    /// assert_eq!(string.capacity(), 16);
+    /// ```
+    pub fn new() -> ArrayString<A> {
+        unsafe {
+            ArrayString {
+                xs: new_array(),
+                len: Index::from(0),
+            }
+        }
+    }
+
+    /// Return the capacity of the `ArrayString`.
+    ///
+    /// ```
+    /// use arrayvec::ArrayString;
+    ///
+    /// let string = ArrayString::<[_; 3]>::new();
+    /// assert_eq!(string.capacity(), 3);
+    /// ```
+    #[inline]
+    pub fn capacity(&self) -> usize { A::capacity() }
+
+    /// Adds the given char to the end of the string.
+    ///
+    /// Returns `None` if the push succeeds, or and returns `Some(c)` if the
+    /// backing array is not large enough to fit the additional char.
+    ///
+    /// ```
+    /// use arrayvec::ArrayString;
+    ///
+    /// let mut string = ArrayString::<[_; 2]>::new();
+    ///
+    /// let none1 = string.push('a');
+    /// let none2 = string.push('b');
+    /// let overflow = string.push('c');
+    ///
+    /// assert_eq!(&string[..], "ab");
+    /// assert_eq!(none1, None);
+    /// assert_eq!(none2, None);
+    /// assert_eq!(overflow, Some('c'));
+    /// ```
+    pub fn push(&mut self, c: char) -> Option<char> {
+        use std::fmt::Write;
+        self.write_char(c).err().map(|_| c)
+    }
+
+    /// Adds the given string slice to the end of the string.
+    ///
+    /// Returns `None` if the push succeeds, or and returns `Some(s)` if the
+    /// backing array is not large enough to fit the string.
+    ///
+    /// ```
+    /// use arrayvec::ArrayString;
+    ///
+    /// let mut string = ArrayString::<[_; 2]>::new();
+    ///
+    /// let none1 = string.push_str("a");
+    /// let overflow1 = string.push_str("bc");
+    /// let none2 = string.push_str("d");
+    /// let overflow2 = string.push_str("ef");
+    ///
+    /// assert_eq!(&string[..], "ad");
+    /// assert_eq!(none1, None);
+    /// assert_eq!(none2, None);
+    /// assert_eq!(overflow1, Some("bc"));
+    /// assert_eq!(overflow2, Some("ef"));
+    /// ```
+    pub fn push_str<'a>(&mut self, s: &'a str) -> Option<&'a str> {
```

Comment:
I don't know, I consider silently failing as a really bad thing. Currently my code using `Array*` looks like `assert!(a.push(..).is_none());`, but if I forget this check, no warning will be issued. This would be different if `push` returned a result.

You can still use it as a "drop-in" for `String` if you really want to, and aren't afraid of seeing a few warnings (or silencing them).